crates/assistant_tools/src/contents_tool.rs

```diff
@@ -0,0 +1,239 @@
+                        if entry.is_dir() {
+                            entry.path.to_str()
+                        } else {
+                            None
+                        }
+                    })
+                })
+                .collect::<Vec<_>>()
+                .join("\n");
+
+            return Task::ready(Ok(output));
+        }
+
+        let Some(project_path) = project.read(cx).find_project_path(&input.path, cx) else {
+            return Task::ready(Err(anyhow!("Path {} not found in project", &input.path)));
+        };
+
+        let Some(worktree) = project
+            .read(cx)
+            .worktree_for_id(project_path.worktree_id, cx)
+        else {
+            return Task::ready(Err(anyhow!("Worktree not found")));
+        };
+        let worktree = worktree.read(cx);
+
+        let Some(entry) = worktree.entry_for_path(&project_path.path) else {
+            return Task::ready(Err(anyhow!("Path not found: {}", input.path)));
+        };
+
+        // If it's a directory, list its contents
+        if entry.is_dir() {
+            let mut output = String::new();
+            let start_index = input
+                .start
+                .map(|line| (line as usize).saturating_sub(1))
+                .unwrap_or(0);
+            let end_index = input
+                .end
+                .map(|line| (line as usize).saturating_sub(1))
+                .unwrap_or(MAX_DIR_ENTRIES);
+            let mut skipped = 0;
+
+            for (index, entry) in worktree.child_entries(&project_path.path).enumerate() {
+                if index >= start_index && index <= end_index {
+                    writeln!(
+                        output,
+                        "{}",
+                        Path::new(worktree.root_name()).join(&entry.path).display(),
+                    )
+                    .unwrap();
+                } else {
+                    skipped += 1;
+                }
+            }
+
+            if output.is_empty() {
+                output.push_str(&input.path);
+                output.push_str(" is empty.");
+            }
+
+            if skipped > 0 {
+                write!(
+                    output,
+                    "\n\nNote: Skipped {skipped} entries. Adjust start and end to see other entries.",
+                ).ok();
+            }
+
+            Task::ready(Ok(output))
+        } else {
+            // It's a file, so read its contents
+            let file_path = input.path.clone();
+            cx.spawn(async move |cx| {
+                let buffer = cx
+                    .update(|cx| {
+                        project.update(cx, |project, cx| project.open_buffer(project_path, cx))
+                    })?
+                    .await?;
+
+                if input.start.is_some() || input.end.is_some() {
+                    let result = buffer.read_with(cx, |buffer, _cx| {
+                        let text = buffer.text();
```

Comment:
Will this perform some kind of "normalization"? By normalization I mean:
- trim trailing whitespace,
- convert all line endings to `\n`,
- convert into a fixed encoding such as UTF-8.